orchagent/bufferorch.cpp

```diff
@@ -416,14 +416,17 @@ task_process_status BufferOrch::processBufferPool(KeyOpFieldsValuesTuple &tuple)
             return task_process_status::task_need_retry;
         }
 
-        sai_status = sai_buffer_api->remove_buffer_pool(sai_object);
-        if (SAI_STATUS_SUCCESS != sai_status)
+        if (SAI_NULL_OBJECT_ID == sai_object)
         {
-            SWSS_LOG_ERROR("Failed to remove buffer pool %s with type %s, rv:%d", object_name.c_str(), map_type_name.c_str(), sai_status);
-            task_process_status handle_status = handleSaiRemoveStatus(SAI_API_BUFFER, sai_status);
-            if (handle_status != task_process_status::task_success)
+            sai_status = sai_buffer_api->remove_buffer_pool(sai_object);
+            if (SAI_STATUS_SUCCESS != sai_status)
             {
-                return handle_status;
+                SWSS_LOG_ERROR("Failed to remove buffer pool %s with type %s, rv:%d", object_name.c_str(), map_type_name.c_str(), sai_status);
+                task_process_status handle_status = handleSaiRemoveStatus(SAI_API_BUFFER, sai_status);
+                if (handle_status != task_process_status::task_success)
+                {
+                    return handle_status;
+                }
             }
         }
         SWSS_LOG_NOTICE("Removed buffer pool %s with type %s", object_name.c_str(), map_type_name.c_str());
```

Comment:
Shouldn't L432-435 be also executed only if the object exists?

Reply:
I tend to keep it as it is.
- `SAI_NULL_OBJECT_ID != sai_object` guarantees the object has been created by SAI, thus unnecessary to call the remove function
- it doesn't guarantee the object hasn't been tracked/recorded in `m_buffer_type_maps`. eg. the SAI returns success with a NULL OBJ ID (even very unlikely to happen)
- we use a map to track reference, which is idempotent and harmless to remove an unexisting reference.
- on the contrary, if we don't remove it from the map when sai_object is NULL OBJ ID, there can be a reference leak in case the SAI returns success with a NULL OBJ ID (again, very unlikely to happen)

Meanwhile, I don't have a strong opinion on it.